### PR TITLE
Add endpoint to create task

### PR DESCRIPTION
**Pull request by pbochis**

### CODE REVIEW

**Comment by flowlo** [comment]
Please commit with an author field that GitHub recognizes.

**Comment by pbochis** [comment]
ready

File: controllers/task.go [hunk 1]
```diff
@@ -79,17 +79,13 @@ func TaskByKey(ctx context.Context, w http.ResponseWriter, r *http.Request) (sta
...
-
 	p, ok := passenger.FromContext(ctx)
 	if !ok {
 		return http.StatusUnauthorized, nil
 	}
 
 	var u model.User
-	if err = datastore.Get(ctx, p.User, &u); err != nil {
+	if err := datastore.Get(ctx, p.User, &u); err != nil {
 		return http.StatusInternalServerError, nil
```
Comment (flowlo) [inline controllers/task.go]: `return http.StatusInternalServerError, err`

File: controllers/code.go [hunk 2]
```diff
@@ -1,24 +1,43 @@
...
 
-	// FIXME(victorbalan): Send correct Content-Type
-	w.Header().Set("Content-Type", "application/octet-stream")
-	w.Header().Set("Content-Disposition", "attachment; filename='template.java'")
+	rc, err := storage.NewReader(util.CloudContext(ctx), util.TemplateBucket, templateName+"/"+fileName)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	content, err := ioutil.ReadAll(rc)
```
Comment (flowlo) [inline controllers/code.go]: Reading from GCS every single time is not fast. Please use memcache. It is plausible that we can use temporal locality when a task is worked on by many coders at the same time.

File: controllers/code.go [hunk 3]
```diff
@@ -1,24 +1,43 @@
...
+	language := mux.Vars(r)["language"]
+	fileName, ok := util.FileNames[language]
+
+	if !ok {
+		return http.StatusInternalServerError, errors.New("language unknown")
+	}
 
-	// FIXME(victorbalan): Send correct Content-Type
-	w.Header().Set("Content-Type", "application/octet-stream")
-	w.Header().Set("Content-Disposition", "attachment; filename='template.java'")
```
Comment (flowlo) [inline controllers/code.go]: Why did you delete those headers (including a FIXME!)?

File: controllers/code.go [hunk 4]
```diff
@@ -1,24 +1,43 @@
...
-func Template(c context.Context, w http.ResponseWriter, r *http.Request) (int, error) {
+func Template(ctx context.Context, w http.ResponseWriter, r *http.Request) (int, error) {
 	if r.Method != "GET" {
 		return http.StatusMethodNotAllowed, nil
 	}
 
-	// TODO(victorbalan): Serve correct template using passenger and result key.
+	templateName := mux.Vars(r)["name"]
+	language := mux.Vars(r)["language"]
+	fileName, ok := util.FileNames[language]
```
Comment (flowlo) [inline controllers/code.go]: Name and language is not specific enough. We'll have to provide a template per-task.

File: controllers/submission.go [hunk 5]
```diff
@@ -141,12 +113,12 @@ func store(ctx context.Context, key *datastore.Key, code, language string) (mode
...
 
 	// We'll be storing code, so check what name the file should have in GCS.
-	fn, ok := fileNames[language]
+	fn, ok := util.FileNames[language]
 	if !ok {
 		return o, errors.New("language unknown")
 	}
 
-	submissionBucket := "coduno"
+	submissionBucket := util.SubmissionBucket
```
Comment (flowlo) [inline controllers/submission.go]: This is inconsistent. Two lines later, `submissionBucket` is changed for development again. If you move the bucket names, you should move all the logic with it.

File: controllers/code.go (hunk 3 above)
Comment (pbochis) [inline controllers/code.go]: Because imo it doesn't make sense to serve an entire file. We only actually need the contents of the file to be displayed in the code editor.

File: controllers/code.go (hunk 4 above)
Comment (pbochis) [inline controllers/code.go]: that temmplateName is task.Template. How would you make it more specific?

Comment (flowlo) [inline controllers/code.go]: OK, I overlooked that.

File: controllers/code.go (hunk 3 above)
Comment (flowlo) [inline controllers/code.go]: Ignore the code editor. This is an API endpoint to download the template. So we should set reasonable headers. If you are too lazy to implement this now, you can leave it as-is, but it does not justify just deleting a FIXME without comment.

File: controllers/code.go [hunk 6]
```diff
@@ -1,24 +1,83 @@
...
+	}
+	fn := templateName + "/" + fileName
+	var content []byte
+	var err error
+	if content, err = fromCache(ctx, fn); err != nil {
+		if content, err = fromStorage(ctx, fn); err != nil {
+			return http.StatusInternalServerError, err
+		}
+	}
+	content, err = ioutil.ReadFile("/home/pbochis/Application.java")
```
Comment (victorbalan) [inline controllers/code.go]: ???

File: controllers/code.go [hunk 7]
```diff
@@ -1,24 +1,83 @@
...
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", "attachment; filename='"+fileName+"'")
+	w.Write([]byte(content))
```
Comment (victorbalan) [inline controllers/code.go]: content is already `[]byte`

File: controllers/task.go [hunk 8]
```diff
@@ -79,18 +79,14 @@ func TaskByKey(ctx context.Context, w http.ResponseWriter, r *http.Request) (sta
...
-
 	p, ok := passenger.FromContext(ctx)
 	if !ok {
 		return http.StatusUnauthorized, nil
 	}
 
 	var u model.User
-	if err = datastore.Get(ctx, p.User, &u); err != nil {
-		return http.StatusInternalServerError, nil
+	if err := datastore.Get(ctx, p.User, &u); err != nil {
```
Comment (victorbalan) [inline controllers/task.go]: remove `:`. It works because `err` is defined as output param

File: controllers/task.go [hunk 9]
```diff
@@ -98,6 +94,31 @@ func Tasks(ctx context.Context, w http.ResponseWriter, r *http.Request) (status
 		return http.StatusUnauthorized, nil
 	}
 
+	if r.Method == "GET" {
```
Comment (victorbalan) [inline controllers/task.go]: I think it would look better with :
```golang
switch r.Method{
    case "GET":
        return getAllTasks()
    case "POST":
        return post()
    default:
        return 401, nil

```

File: controllers/task.go [hunk 10]
```diff
@@ -98,6 +94,31 @@ func Tasks(ctx context.Context, w http.ResponseWriter, r *http.Request) (status
...
+}
+
+func createTask(ctx context.Context, w http.ResponseWriter, r *http.Request) (status int, err error) {
+	var task model.Task
+
+	if err = json.NewDecoder(r.Body).Decode(&task); err != nil {
+		return http.StatusBadRequest, err
+	}
+
+	key, err := task.Put(ctx, nil)
```
Comment (victorbalan) [inline controllers/task.go]: usual way for this is:
```go
var key *datastore.Key
if key, err = task.Put(ctx, nil);err != nil {
    return http.StatusInternalServerError, err
}
```

File: main.go [hunk 11]
```diff
@@ -33,7 +33,7 @@ func main() {
 	r := mux.NewRouter()
 	r.HandleFunc("/subscriptions", hsts(controllers.Subscriptions))
 
-	r.HandleFunc("/code/download", hsts(guard(controllers.Template)))
+	r.HandleFunc("/code/download/{name}/{language}", hsts(guard(controllers.Template)))
```
Comment (victorbalan) [inline main.go]: use `setup(controllers.Template)`

File: controllers/code.go (hunk 6 above)
Comment (victorbalan) [inline controllers/code.go]: https://imgflip.com/i/qs53p

Comment (flowlo) [inline controllers/code.go]: @victorbalan I nearly shit my pants on this one! :laughing:
